### ISSUE TITLE
Crossentropy 종류

### ISSUE THREAD

**Issue by CheezEun** [CheezEun/2022_Google_ML_Bootcamp#8]
TF Developers 과정 Course2 - Week 4 (Multiclass 분류) 과제에서 모델의 output은 ``(None, 26)`` 인데 반해 label은 ``(None, 1)`` 으로 서로 다른 형태를 가지는 데서 생기는 오류 때문에 막혀서 한참 헤맸다. 

결론적으로 말하자면 loss의 선택을 잘못했기 때문에 발생했다.
그럼 어떤 상황에서 어떤 loss를 사용해야 할까?

1. ``loss = 'binary_crossentropy'``
  지금까지 해왔던 binary classification ( True or False 분류 문제)에서 사용하며, 다음과 같은 상황에서 사용한다.
    - Dense layer
    - acitvation = 'sigmoid'
    - output units 개수 = 1 또는 n_class
    - y_label 의 구조 = (None, n_class)

    e.g. 
    $\quad$ cats vs. dogs $\rightarrow$ n_class = 1 인 이진 분류
    $\quad$ 블로그 게시물의 태그 $\rightarrow$ 각각의 태그(#Coursera 등) 마다 0 또는 1을 가지므로 n_class = (태그 종류) 인 이진 분류

<br>

2. ``loss = 'categorical_crossentropy``
  Multi-class classification 에서 사용하며, 다음과 같은 상황에서 사용한다.
    - Dense layer
    - activation = 'softmax'
    - output units 개수 = n_class
    - y_label의 구조 = (None, n_class)
    - 즉, **label이 one-hot vector인 경우** (e.g. [0, 1, 0, 0, 0] )

<br>

3. ``loss = 'sparse_categorical_crossentropy``
  Multi-class classification 에서 사용하며, 다음과 같은 상황에서 사용한다.
    - Dense layer
    - activation = 'softmax'
    - output units 개수 = n_class
    - y_label의 구조 = (None, )
    - 즉, **label이 integer인 경우** (e.g. 포메라니안 = 1, 불독 = 2, 치와와 = 3, ... )

참고 문서 : [Reference](https://cheris8.github.io/artificial%20intelligence/DL-Keras-Loss-Function/)
